pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	utilerrors "k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	sentinelFileName = "..sentinel"
+	tsFormat         = "1981_02_01_15_04_05"
+)
+
+func ts() string {
+	return time.Now().Format(tsFormat)
+}
+
+// TODO: inject coordinates for logging traceability to pod/volume
+
+// AtomicWriter handles writing a payload to a target directory atomically.
+type AtomicWriter struct {
+	TargetDir       string
+	IdentifierToken string
+	LogContext      string
+}
+
+// currentStateDirName returns the name of the directory within the writer's
+// target directory that should hold the current projected state.
+func (w *AtomicWriter) currentStateDirName() string {
+	return fmt.Sprintf("..%v", w.IdentifierToken)
+}
+
+// updateStateDirName returns the name
+func (w *AtomicWriter) updateStateDirName() string {
+	return fmt.Sprintf("..%v_tmp", w.IdentifierToken)
+}
+
+// Write ensures that the given payload is projected onto disk, updating the
+// contents of the AtomicWriter's target directory atomically if any writes
+// need to be performed.  A sentinel file named "..sentinel" in the target
+// directory is updated, allowing consumers of the directory to listen for
+// updates to the projected data by monitoring the sentinel file with inotify
+// or fanotify.
+//
+// The visible files in this volume are symlinks to files in the writer's
+// current state directory directory.  Actual files are stored in a hidden
+// timestamped directory which is symlinked to by the current state directory.
+// The timestamped directory and current state symlink are created in the
+// writer's target dir.  This scheme allows the files to be atomically updated by
+// changing the target of the current state symlink.
```

Comment:
You should document that, to make this work, all directory entries that start with ".." are reserved.  This needs to be very visible to consumers of this lib

You should also document behavior around adding/removing files and dirs across updates.